### PR TITLE
implement both C indirect-function and lisp one

### LINKED ISSUE #332: Port indirect-function to Rust
```c
/* If OBJECT is a symbol, find the end of its function chain and
   return the value found there.  If OBJECT is not a symbol, just
   return it.  If there is a cycle in the function chain, signal a
   cyclic-function-indirection error.

   This is like Findirect_function, except that it doesn't signal an
   error if the chain ends up unbound.  */
Lisp_Object
indirect_function (register Lisp_Object object)
{
  Lisp_Object tortoise, hare;

  hare = tortoise = object;

  for (;;)
    {
      if (!SYMBOLP (hare) || NILP (hare))
	break;
      hare = XSYMBOL (hare)->function;
      if (!SYMBOLP (hare) || NILP (hare))
	break;
      hare = XSYMBOL (hare)->function;

      tortoise = XSYMBOL (tortoise)->function;

      if (EQ (hare, tortoise))
	xsignal1 (Qcyclic_function_indirection, object);
    }

  return hare;
}

DEFUN ("indirect-function", Findirect_function, Sindirect_function, 1, 2, 0,
       doc: /* Return the function at the end of OBJECT's function chain.
If OBJECT is not a symbol, just return it.  Otherwise, follow all
function indirections to find the final function binding and return it.
Signal a cyclic-function-indirection error if there is a loop in the
function chain of symbols.  */)
  (register Lisp_Object object, Lisp_Object noerror)
{
  Lisp_Object result;

  /* Optimize for no indirection.  */
  result = object;
  if (SYMBOLP (result) && !NILP (result)
      && (result = XSYMBOL (result)->function, SYMBOLP (result)))
    result = indirect_function (result);
  if (!NILP (result))
    return result;

  return Qnil;
}
```

### PR DESCRIPTION
**Pull request by Yarikx**
closes #332

### CODE REVIEW

File: src/data.c [hunk 1]
```diff
@@ -1806,61 +1806,6 @@ selected frame's terminal device).  */)
...
-       doc: /* Return the function at the end of OBJECT's function chain.
-If OBJECT is not a symbol, just return it.  Otherwise, follow all
-function indirections to find the final function binding and return it.
-Signal a cyclic-function-indirection error if there is a loop in the
-function chain of symbols.  */)
-  (register Lisp_Object object, Lisp_Object noerror)
-{
-  Lisp_Object result;
-
-  /* Optimize for no indirection.  */
```
Comment (Wilfred) [inline src/data.c]: Do you think this comment is worth preserving?

Comment (Yarikx) [reply src/data.c]: Not sure, but ok. Will add to rust code as well.

File: rust_src/src/data.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+use remacs_macros::lisp_fn;
```
Comment (brotzeit) [inline rust_src/src/data.rs]: "//! Primitive operations on Lisp data types for GNU Emacs Lisp interpreter." is missing here.

Comment (Yarikx) [reply rust_src/src/data.rs]: Good point!

**Comment by birkenfeld** [comment]
Thanks!

**Pull request merged**